test/e2e/e2e_test.go

```diff
@@ -0,0 +1,409 @@
+
+It supports dynamically creating (and post-test deleting) a DOKS cluster to run a driver-under-test in. The environment
+variable DIGITALOCAN_ACCESS_TOKEN must be set to a DigitalOcean API key for this purpose.
+The cluster will be tagged with "csi-e2e-test" and "version:<sanitized Kubernetes version>" where
+<sanitized Kubernetes version> is the Kubernetes major-minor version replacing dots by dashes, e.g., "version:1-16"
+(DigitalOcean tags must not contain dots).
+The name of a cluster will be "csi-e2e-<sanitized Kubernetes version>-test-<suffix>" where <suffix> is a random
+alphanumeric suffix if not customized through the corresponding command-line flag.
+
+One or more Kubernetes major-minor versions to run tests for may be given. If omitted, tests will be run for all supported
+Kubernetes versions.
+
+External storage end-to-end tests require a Kubernetes version-specific testdriver YAML file to be defined. An error is
+returned if no corresponding file is found for a given Kubernetes release.
+
+Examples:
+
+# Run tests for all supported versions:
+e2e.test
+
+# Run tests for 1.16 only:
+e2e.test 1.16
+
+# Run tests for 1.16 and 1.14 (but not 1.15):
+e2e.test 1.16 1.14
+
+# Create cluster with a specific suffix:
+e2e.test --name-suffix=$(git rev-parse --short HEAD)
+
+# Retain cluster after completion of the tests:
+e2e.test --retain
+
+# Use cluster referenced by kubeconfig file instead of using a dynamic cluster:
+e2e.test --kubeconfig=$HOME/.kube/config
+
+# Use a custom driver image:
+e2e.test --driver-image=timoreimann/do-csi-plugin:dev
+
+# Use a custom end-to-end test runner image:
+e2e.test --runner-image=timoreimann/k8s-e2e-test-runner:latest
+
+Options:`)
+		flag.PrintDefaults()
+	}
+}
+
+func TestMain(m *testing.M) {
+	flag.BoolVar(&p.long, "long", false, "Run long tests")
+	flag.StringVar(&p.driverImage, "driver-image", "", "The driver container image to use. Triggers a deployment of the \"latest\"-suffixed development manifest into the cluster if given. Otherwise, the built-in driver of the cluster is used.")
+	flag.StringVar(&p.runnerImage, "runner-image", testRunnerImage, "The end-to-end runner image to use.")
+	flag.StringVar(&p.focus, "focus", "", "A custom ginkgo focus to use for external storage tests. Defaults to running all external tests.")
+	flag.StringVar(&p.kubeconfig, "kubeconfig", "", "The kubeconfig file to use. For DOKS clusters where the kubeconfig has been retrieved via doctl, the DIGITALOCEAN_ACCESS_TOKEN environment variable must be set. If not specified, add-hoc DOKS clusters will be created and cleaned up afterwards for each tested Kubernetes version (unless -retain is also specified).")
+	flag.StringVar(&p.nameSuffix, "name-suffix", "", "A suffix to append to the cluster name. If not specified, a random suffix will be chosen. Ignored if -kubeconfig is specified.")
+	flag.BoolVar(&p.retainClusters, "retain", false, "Retain/keep the created cluster(s). Ignored if -kubeconfig is specified.")
```

Comment:
This flag may make more sense for failed clusters specifically vs all clusters. Theoretically if the test run succeeds then it should just be a blank state cluster after test cleanup which isn't all that useful imo.

Reply:
That's an excellent point. I'll make the necessary adjustments.

Reply:
Pushed another commit to implement the suggested change in behavior.